openpype/hosts/max/api/preview_animation.py

```diff
@@ -0,0 +1,309 @@
+        renderRatio = float(dib_width / dib_height)
+        if viewportRatio <= renderRatio:
+            heightCrop = (dib_width / renderRatio)
+            topEdge = int((dib_height - heightCrop) / 2.0)
+            tempImage_bmp = rt.bitmap(dib_width, heightCrop)
+            src_box_value = rt.Box2(0, topEdge, dib_width, heightCrop)
+        else:
+            widthCrop = dib_height * renderRatio
+            leftEdge = int((dib_width - widthCrop) / 2.0)
+            tempImage_bmp = rt.bitmap(widthCrop, dib_height)
+            src_box_value = rt.Box2(0, leftEdge, dib_width, dib_height)
+        rt.pasteBitmap(dib, tempImage_bmp, src_box_value, rt.Point2(0, 0))
+
+        # copy the bitmap and close it
+        rt.copy(tempImage_bmp, preview_res)
+        rt.close(tempImage_bmp)
+
+        rt.save(preview_res)
+        rt.close(preview_res)
+
+        rt.close(dib)
+
+        if rt.keyboard.escPressed:
+            rt.exit()
+    # clean up the cache
+    rt.gc(delayed=True)
+
+
+def render_preview_animation(
+        instance, staging_dir,
+        ext, review_camera,
+        startFrame=None, endFrame=None,
+        resolution=None,
+        viewport_options=None):
+    """Render camera review animation
+
+    Args:
+        instance (pyblish.api.instance): Instance
+        filepath (str): filepath
+        review_camera (str): viewport camera for preview render
+        startFrame (int): start frame
+        endFrame (int): end frame
+        viewport_options (dict): viewport setting options
+    """
+
+    if startFrame is None:
+        startFrame = int(rt.animationRange.start)
+    if endFrame is None:
+        endFrame = int(rt.animationRange.end)
+    if viewport_options is None:
+        viewport_options = viewport_options_for_preview_animation()
+    if resolution is None:
+        resolution = (1920, 1080)
+    with play_preview_when_done(False):
+        with viewport_camera(review_camera):
+            width, height = resolution
+            with render_resolution(width, height):
+                if int(get_max_version()) < 2024:
+                    with viewport_preference_setting(
+                            viewport_options["general_viewport"],
+                            viewport_options["nitrous_viewport"],
+                            viewport_options["vp_btn_mgr"]):
+                        percentSize = viewport_options.get("percentSize", 100)
+
+                        publish_preview_sequences(
+                            staging_dir, instance.name,
+                            startFrame, endFrame, percentSize, ext)
+                else:
+                    fps = instance.data["fps"]
+                    rt.completeRedraw()
+                    preview_arg = publish_review_animation(
+                        instance, staging_dir,
+                        startFrame, endFrame,
+                        ext, fps, viewport_options)
+                    rt.execute(preview_arg)
+
+    rt.completeRedraw()
```

Comment:
I still feel like this code can have a much easier entry point. Consider something like this:

```suggestion
def _render_preview_animation_max_2024(filepath, start,
                             end, ext, fps, viewport_options):
    """Function to set up preview arguments in MaxScript.
    ****For 3dsMax 2024+

    Args:
        filepath (str): filepath for render output without frame number and
            extension, for example: /path/to/file
        start (int): startFrame
        end (int): endFrame
        fps (float): fps value
        viewport_options (dict): viewport setting options, e.g.
            {"vpStyle": "defaultshading", "vpPreset": "highquality"}

    Returns:
        list: Created files
    """

    filepath = filepath.replace("\\", "/")
    filepath = f"{filepath}..{ext}"
    frame_template = f"{filepath}.{{:04d}}.{ext}"

    job_args = list()
    default_option = f'CreatePreview filename:"{filepath}"'
    job_args.append(default_option)
    frame_option = f"outputAVI:false start:{start} end:{end} fps:{fps}" # noqa
    job_args.append(frame_option)

    for key, value in viewport_options.items():
        if isinstance(value, bool):
            if value:
                job_args.append(f"{key}:{value}")

        elif isinstance(value, str):
            if key == "vpStyle":
                if value == "Realistic":
                    value = "defaultshading"
                elif value == "Shaded":
                    log.warning(
                        "'Shaded' Mode not supported in "
                        "preview animation in Max 2024.\n"
                        "Using 'defaultshading' instead.")
                    value = "defaultshading"
                elif value == "ConsistentColors":
                    value = "flatcolor"
                else:
                    value = value.lower()
            elif key == "vpPreset":
                if value == "Quality":
                    value = "highquality"
                elif value == "Customize":
                    value = "userdefined"
                else:
                    value = value.lower()
            job_args.append(f"{key}: #{value}")

    auto_play_option = "autoPlay:false"
    job_args.append(auto_play_option)

    job_str = " ".join(job_args)
    log.debug(job_str)

    rt.completeRedraw()
    rt.execute(job_str)

    # Return the created files
    return [frame_template.format(frame) for frame in range(start, end + 1)]


def _render_preview_animation_max_pre_2024(
        filepath, startFrame, endFrame, percentSize, ext):
    """Render viewport animation by creating bitmaps

    ***For 3dsMax Version <2024

    Args:
        filepath (str): filepath without frame numbers and extension
        startFrame (int): start frame
        endFrame (int): end frame
        percentSize (int): percentage of the resolution
        ext (str): image extension

    Returns:
        list: Created filepaths

    """
    # get the screenshot
    resolution_percentage = float(percentSize) / 100
    res_width = rt.renderWidth * resolution_percentage
    res_height = rt.renderHeight * resolution_percentage

    viewportRatio = float(res_width / res_height)

    frame_template = "{}.{{:04}}.{}".format(filepath, ext)
    frame_template.replace("\\", "/")

    files = []

    user_cancelled = False
    for frame in range(startFrame, endFrame + 1):
        rt.sliderTime = frame

        filepath = frame_template.format(frame)
        preview_res = rt.bitmap(
            res_width, res_height, filename=filepath
        )
        dib = rt.gw.getViewportDib()
        dib_width = float(dib.width)
        dib_height = float(dib.height)
        renderRatio = float(dib_width / dib_height)
        if viewportRatio <= renderRatio:
            heightCrop = (dib_width / renderRatio)
            topEdge = int((dib_height - heightCrop) / 2.0)
            tempImage_bmp = rt.bitmap(dib_width, heightCrop)
            src_box_value = rt.Box2(0, topEdge, dib_width, heightCrop)
        else:
            widthCrop = dib_height * renderRatio
            leftEdge = int((dib_width - widthCrop) / 2.0)
            tempImage_bmp = rt.bitmap(widthCrop, dib_height)
            src_box_value = rt.Box2(0, leftEdge, dib_width, dib_height)
        rt.pasteBitmap(dib, tempImage_bmp, src_box_value, rt.Point2(0, 0))

        # copy the bitmap and close it
        rt.copy(tempImage_bmp, preview_res)
        rt.close(tempImage_bmp)

        rt.save(preview_res)
        rt.close(preview_res)

        rt.close(dib)

        files.append(filepath)

        if rt.keyboard.escPressed:
            rt.exit()
            user_cancelled = True
            break

    # clean up the cache
    rt.gc(delayed=True)

    if user_cancelled:
        raise RuntimeError("User cancelled rendering of viewport animation.")

    return files


def render_preview_animation(
        filepath,
        ext, 
        review_camera,
        start_frame=None, 
        end_frame=None,
        width=1920,
        height=1080,
        fps=None,
        viewport_options=None):
    """Render camera review animation

    Args:
        filepath (str): filepath to render to, without frame number and 
            extension
        ext (str): output file extension
        review_camera (str): viewport camera for preview render
        start_frame (int): start frame
        end_frame (int): end frame
        width (int): render resolution width
        height (int): render resolution height
        fps (float): frames per second
        viewport_options (dict): viewport setting options

    Returns:
        list: Rendered output files

    """

    if start_frame is None:
        start_frame = int(rt.animationRange.start)
    if end_frame is None:
        end_frame = int(rt.animationRange.end)
        
    if viewport_options is None:
        viewport_options = viewport_options_for_preview_animation()

    with play_preview_when_done(False):
        with viewport_camera(review_camera):
            with render_resolution(width, height):
                if int(get_max_version()) < 2024:
                    with viewport_preference_setting(
                            viewport_options["general_viewport"],
                            viewport_options["nitrous_viewport"],
                            viewport_options["vp_btn_mgr"]
                    ):
                        percentSize = viewport_options.get("percentSize", 100)

                        return _render_preview_animation_max_pre_2024(
                            filepath,
                            start_frame,
                            end_frame,
                            percentSize,
                            ext
                        )
                else:
                    return _render_preview_animation_max_2024(
                        filepath,
                        start_frame, 
                        end_frame,
                        ext, 
                        fps, 
                        viewport_options
                    )
```

Reply:
it works without `rt.exit()`